torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -200,55 +199,55 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
+      autogradContext->retrieveGraphTask(), graphRoot, /*async_mode=*/true);
 
   // Build a future that waits for the callbacks to execute (since callbacks
   // execute after the original future is completed). This ensures we return a
   // future that waits for all gradient accumulation to finish.
   auto accumulateGradFuture = std::make_shared<rpc::FutureMessage>();
 
-  futureGrads->addCallback(
-      [autogradContext, outputEdges, accumulateGradFuture](
-          const variable_list& grads,
-          const c10::optional<torch::utils::FutureError>& error) {
-        if (error) {
-          // Don't accumulate gradients if we receive an error.
-          // We must add the node information here since DistEngine::execute
-          // waits on accumulateGradFuture and will throw an exception once we
-          // set the error below.
-          std::string errorMsg = c10::str(
-              "Error on Node ",
-              DistAutogradContainer::getInstance().getWorkerId(),
-              ": ",
-              error->what());
-          accumulateGradFuture->setError(errorMsg);
-          return;
-        }
-
-        try {
-          TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
-
-          // Accumulate all the gradients in the context.
-          for (size_t i = 0; i < grads.size(); i++) {
-            // It is possible that the grad is not defined since a separate
-            // invocation of the autograd engine on the same node might actually
-            // compute this gradient. Also accumulate grads only for
-            // AccumulateGrad function.
-            if (grads[i].defined() &&
-                dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
-              auto& variable = std::static_pointer_cast<AccumulateGrad>(
-                                   outputEdges[i].function)
-                                   ->variable;
-              autogradContext->accumulateGrad(
-                  variable, grads[i], 1 /* num_expected_refs */);
-            }
-          }
+  futureGrads->addCallback([autogradContext, outputEdges, accumulateGradFuture](
+                               const FutureVariableList& futureGrads) {
+    if (futureGrads.hasError()) {
+      // Don't accumulate gradients if we receive an error.
+      // We must add the node information here since DistEngine::execute
+      // waits on accumulateGradFuture and will throw an exception once we
+      // set the error below.
+      std::string errorMsg = c10::str(
+          "Error on Node ",
+          DistAutogradContainer::getInstance().getWorkerId(),
+          ": ",
+          futureGrads.error()->what());
+      accumulateGradFuture->setError(errorMsg);
+      return;
+    }
 
-          accumulateGradFuture->markCompleted(rpc::Message());
-        } catch (std::exception& e) {
-          accumulateGradFuture->setErrorIfNeeded(e.what());
+    try {
+      const variable_list& grads = futureGrads.constValue();
```

Comment:
Just realized that if we move the value out of futureGrads here, we don't need the use_count assert.

Reply:
cc @wanchaol since this is related to the autograd PR.

Reply:
Cool! does this means that we don't need to return a share_ptr& for future_results?